In [1]:
spark

Intitializing Scala interpreter ...

Spark Web UI available at http://10.80.102.113:4040
SparkContext available as 'sc' (version = 2.3.1, master = local[*], app id = local-1543604048782)
SparkSession available as 'spark'


res0: org.apache.spark.sql.SparkSession = org.apache.spark.sql.SparkSession@165ab729


In [2]:
spark.version

res1: String = 2.3.1


In [18]:
val sqlContext = new org.apache.spark.sql.SQLContext(sc)

sqlContext: org.apache.spark.sql.SQLContext = org.apache.spark.sql.SQLContext@7cdaf308


# Kaggle Dataset Load

# DoE Dataset Load concise

In [3]:
val inHeader = sc.textFile("MERGED2016_17_PP.csv").map(x => x + ",YEAR").map(line =>line.split(","))

val in2016 = sc.textFile("MERGED2016_17_PP.csv").map(x => x + ",2016").map(line =>line.split(","))

val in2015 = sc.textFile("MERGED2015_16_PP.csv").map(x => x + ",2015").map(line =>line.split(","))

val in2014 = sc.textFile("MERGED2014_15_PP.csv").map(x => x + ",2014").map(line =>line.split(","))

val in2013 = sc.textFile("MERGED2013_14_PP.csv").map(x => x + ",2013").map(line =>line.split(","))

val in2012 = sc.textFile("MERGED2012_13_PP.csv").map(x => x + ",2012").map(line =>line.split(","))

val in2011 = sc.textFile("MERGED2011_12_PP.csv").map(x => x + ",2011").map(line =>line.split(","))

val in2010 = sc.textFile("MERGED2010_11_PP.csv").map(x => x + ",2010").map(line =>line.split(","))

val in2009 = sc.textFile("MERGED2009_10_PP.csv").map(x => x + ",2009").map(line =>line.split(","))

val in2008 = sc.textFile("MERGED2008_09_PP.csv").map(x => x + ",2008").map(line =>line.split(","))

val in2007 = sc.textFile("MERGED2007_08_PP.csv").map(x => x + ",2007").map(line =>line.split(","))


inHeader: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[3] at map at <console>:25
in2016: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[7] at map at <console>:27
in2015: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[11] at map at <console>:29
in2014: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[15] at map at <console>:31
in2013: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[19] at map at <console>:33
in2012: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[23] at map at <console>:35
in2011: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[27] at map at <console>:37
in2010: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[31] at map at <console>:39
in2009: org.apache.spark.rdd.RDD[Array[String...

## Compile the 10 year RDD.

In [11]:
val scoreCard10Year = sc.union(in2016, in2015, in2014, in2013, in2012, in2011, in2010, in2009, in2008, in2007)
val twoYear = in2016.union(in2015)
scoreCard10Year.take(1)
twoYear.take(1)

scoreCard10Year: org.apache.spark.rdd.RDD[Array[String]] = UnionRDD[59] at union at <console>:40
twoYear: org.apache.spark.rdd.RDD[Array[String]] = UnionRDD[60] at union at <console>:41
res9: Array[Array[String]] = Array(Array(UNITID, OPEID, OPEID6, INSTNM, CITY, STABBR, ZIP, ACCREDAGENCY, INSTURL, NPCURL, SCH_DEG, HCM2, MAIN, NUMBRANCH, PREDDEG, HIGHDEG, CONTROL, ST_FIPS, REGION, LOCALE, LOCALE2, LATITUDE, LONGITUDE, CCBASIC, CCUGPROF, CCSIZSET, HBCU, PBI, ANNHI, TRIBAL, AANAPII, HSI, NANTI, MENONLY, WOMENONLY, RELAFFIL, ADM_RATE, ADM_RATE_ALL, SATVR25, SATVR75, SATMT25, SATMT75, SATWR25, SATWR75, SATVRMID, SATMTMID, SATWRMID, ACTCM25, ACTCM75, ACTEN25, ACTEN75, ACTMT25, ACTMT75, ACTWR25, ACTWR75, ACTCMMID, ACTENMID, ACTMTMID, ACTWRMID, SAT_AVG, SAT_AVG_ALL, PCIP01, PCIP03, PCIP04, PCI...

# Create Parquet file

Reduced storage from CSV size of 1.5GB to Parquet file size of 294MB using the Snappy compression method. #thatswow

In [13]:
val dataDF = scoreCard10Year.toDF()                 // convert RDD to DataFrame
// dataDF.write.parquet("data.parquet")                // write to parquet once

dataDF: org.apache.spark.sql.DataFrame = [value: array<string>]


## Create a DF that references the Parquet files.

In [19]:
val pqScoreCard10Year = sqlContext.read.parquet("data.parquet")


pqScoreCard10Year: org.apache.spark.sql.DataFrame = [value: array<string>]


# Create a header index of the dataset

In [14]:
val header2016 = sc.parallelize(inHeader.take(1))

val header2016Index = header2016.map(_.mkString(",")).flatMap(line => line.split(",")).zipWithIndex

// header2016Index.coalesce(1).saveAsTextFile("index") //Textfiles only made once


header2016: org.apache.spark.rdd.RDD[Array[String]] = ParallelCollectionRDD[64] at parallelize at <console>:27
header2016Index: org.apache.spark.rdd.RDD[(String, Long)] = ZippedWithIndexRDD[67] at zipWithIndex at <console>:29


## Count of rows for the compiled DoE Scorecard

In [15]:
scoreCard10Year.count()

res11: Long = 74181


In [16]:
dataDF.count()

res12: Long = 74181


### Number of Schools in the DoE dataset

In [8]:
scoreCard10Year.map(x=> x(1)).distinct().count()

res5: Long = 9743


# St Thomas RDD  
This RDD contains all of the St. Thomas results over the years

In [23]:
val ust = scoreCard10Year.filter(x => x(0) contains "174914")
val pqUst = pqScoreCard10Year.filter(x => x(0) contains "174914")

<console>: 43: error: value contains is not a member of Any

In [22]:
ust.take(1)

res14: Array[Array[String]] = Array(Array(174914, 00234500, 002345, University of St Thomas, Saint Paul, MN, 55105-1078, Higher Learning Commission, www.stthomas.edu, www.stthomas.edu/financialaid/undergraduate/resources/netpricecalculator/, NULL, 0, 1, 1, 3, 4, 2, 27, 4, 11, NULL, 44.943439, -93.188986, 17, 10, 13, 0, 0, 0, 0, 0, 0, 0, 0, 0, 30, 0.8266, 0.82655521620318, 500, 660, 550, 630, NULL, NULL, 580, 590, NULL, 24, 29, 23, 29, 24, 28, NULL, NULL, 27, 26, 26, NULL, 1204, 1204, 0, 0.0037, 0, 0.0007, 0.0504, 0, 0.0212, 0, 0.0358, 0.0651, 0, 0.0073, 0.0066, 0, 0.0139, 0.0066, 0, 0.0658, 0.0117, 0, 0.0102, 0, 0.0417, 0, 0.0197, 0, 0.0336, 0, 0.0117, 0.0621, 0, 0, 0, 0, 0.0058, 0.0417, 0.4751, 0.0095, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, ...

In [11]:
ust.count()

res7: Long = 10


In [13]:
val ustHeader = header2016.union(ust)
// ustHeader.map(_.mkString(",")).coalesce(1).saveAsTextFile("ustHeader")

ustHeader: org.apache.spark.rdd.RDD[Array[String]] = UnionRDD[65] at union at <console>:43


# Regional Peers RDD

This RDD filters region 4 (upper midwest/plains), Schools that offer graduate degrees as the highest degree, and a carnegie classification of 4 year, medium size, highly residential. 

In [14]:
val regionalPeers = scoreCard10Year.filter(x => x(18) == "4").filter(x => x(15) == "4").filter(x => x(25) == "13")

regionalPeers: org.apache.spark.rdd.RDD[Array[String]] = MapPartitionsRDD[71] at filter at <console>:38


In [15]:
regionalPeers.count()

res10: Long = 20


### Average SAT score for peers in the region. Ours is 1204

In [38]:
scoreCard10Year.map(x => x(1899)).map(_.mkString("")).take(1) // I verified index 1899 in all of my load RDDs

res31: Array[String] = Array(2016)


In [39]:
regionalPeers.filter(x => x(1899) == "2016").map(x => x(59)).filter(x => x != "NULL").count()


res32: Long = 18


In [40]:
regionalPeers.map(x => x(59)).filter(x => x != "NULL").map(_.toInt).mean()


res33: Double = 1080.7222222222224


# Create an index of the header 

Really had a tough time figuring out how to get the header loaded as just a string. Loading it from the original source results in type RDD[Array[String]], so I ended up using a CSV file that I outputed from the original input. Ended up just needing a **flatMap** to keep it in the correct format. 

# Minnesota Schools  
This rdd contains all of the Minnesota universities for a statewide comparison

In [ ]:
val mnColleges = scoreCard10Year.filter( x => x(5) contains "MN").filter( x => x(1899) contains "2016")

In [ ]:
val mnColLoc = mnColleges.filter(x => x(1899) contains "2016").filter(x => x(36) != "NULL").map(x => x(3) + "," + x(21) + "," + x(22) + "," + x(36))

In [ ]:
// mnColLoc.take(5).foreach(println) // Only needs to run once.

### Create and export RDD to CSV for mapping 

In [ ]:
val mnColLocHead = Array("Name,Lat,Long,AdmRate")
val mnLocRDD = sc.parallelize(mnColLocHead)
val mnColLocFin = mnLocRDD.union(mnColLoc)
//mnColLocFin.coalesce(1).saveAsTextFile("MnSchools") // only run once